test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go

```diff
@@ -0,0 +1,391 @@
+		return errors.Wrapf(err, "failed to create docker machine with name %s", name)
+	}
+	return nil
+}
+
+// reconcileDockerMachines creates and deletes DockerMachines to match the MachinePool's desired number of replicas and infrastructure spec.
+// It is responsible for
+// - Ensuring each Docker container has an associated DockerMachine by creating one if it doesn't already exist.
+// - Ensuring that deletion for Docker container happens by calling delete on the associated Machine so that the node is cordoned/drained and the infrastructure is cleaned up.
+// - Deleting DockerMachines referencing a container whose Kubernetes version or custom image no longer matches the spec.
+// - Deleting DockerMachines that correspond to a deleted/non-existent Docker container.
+// - Deleting DockerMachines when scaling down such that DockerMachines whose owner Machine has the clusterv1.DeleteMachineAnnotation is given priority.
+func (r *DockerMachinePoolReconciler) reconcileDockerMachines(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Reconciling DockerMachines", "dockerMachinePool", dockerMachinePool.Name, "namespace", dockerMachinePool.Namespace)
+
+	dockerMachineList, err := getDockerMachines(ctx, r.Client, *cluster, *machinePool, *dockerMachinePool)
+	if err != nil {
+		return err
+	}
+
+	dockerMachineSet := make(map[string]infrav1.DockerMachine)
+	for _, dockerMachine := range dockerMachineList.Items {
+		dockerMachineSet[dockerMachine.Name] = dockerMachine
+	}
+
+	labelFilters := map[string]string{dockerMachinePoolLabel: dockerMachinePool.Name}
+
+	// List the Docker containers. This corresponds to a InfraMachinePool instance for providers.
+	externalMachines, err := docker.ListMachinesByCluster(ctx, cluster, labelFilters)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all machines in the cluster")
+	}
+
+	// Step 1:
+	// Create a DockerMachine for each Docker container so we surface the information to the user. Use the same name as the Docker container for the Docker Machine for ease of lookup.
+	// Providers should iterate through their infrastructure instances and ensure that each instance has a corresponding InfraMachine.
+	for _, machine := range externalMachines {
+		if _, ok := dockerMachineSet[machine.Name()]; !ok {
+			log.V(2).Info("Creating a new DockerMachine for Docker container", "container", machine.Name())
+			dockerMachine, err := r.createDockerMachine(ctx, machine.Name(), cluster, machinePool, dockerMachinePool)
+			if err != nil {
+				return errors.Wrap(err, "failed to create a new docker machine")
+			}
+
+			dockerMachineSet[dockerMachine.Name] = *dockerMachine
+		} else {
+			log.V(4).Info("DockerMachine already exists, nothing to do", "name", machine.Name())
+		}
+	}
+
+	externalMachineSet := make(map[string]*docker.Machine)
```

Comment:
nit, this is a Map, not a Set

Also, let's move this up after reading external machines (similar for dockerMachineMap)